### PR TITLE
Added test verifying corner cases of specification

### PR DESCRIPTION
**Pull request by honza-kasik**
Added test verifying corner cases of specification

Please make sure your PR meets the following requirements:
- [x] Pull Request contains a description of the changes
- [x] Pull Request does not include fixes for multiple issues/topics
- [x] Code is formatted, imports ordered, code compiles and tests are passing
- [ ] Link to the passing job is provided WIP: Some tests are failing because of a bug.
- [x] Code is self-descriptive and/or documented
- [x] Description of the tests scenarios is included (see #46)

### CODE REVIEW

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/CorruptedJwtTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        final URL privateKeyUrl = CorruptedJwtTest.class.getClassLoader().getResource("pki/key.private.pkcs8.pem");
+        if (privateKeyUrl == null) {
+            throw new IllegalStateException("Private key wasn't found in resources!");
+        }
+
+        final RsaKeyTool keyTool = RsaKeyTool.newKeyTool(privateKeyUrl.toURI());
+
+        final JsonWebToken corruptedJsonJWT = new JwtHelper(keyTool, "issuer").generateJwtJsonCorrupted("FAKE_USER");
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/CorruptedJwtTest.java]: Use static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtIssuerValueTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+     *                the value set to the runtime which will be used for comparision meaning this JWT should be
+     *                rejected.
+     * @tpPassCrit JWT is rejected and user receives 401/Unauthorized.
+     * @tpSince EAP 7.4.0.CD19
+     */
+    @Test
+    public void supplyJwtWithIncorrectIssuerClaimTest(@ArquillianResource URL url) {
+        final JsonWebToken jwt = new JwtHelper(keyTool, "fooqux42").generateProperSignedJwt();
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtIssuerValueTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtIssuerValueTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    /**
+     * @tpTestDetails Supply a JWT with unset issuer claim to the server.
+     * @tpPassCrit JWT is rejected and user receives 401/Unauthorized.
+     * @tpSince EAP 7.4.0.CD19
+     */
+    @Test
+    public void supplyJwtWithUnsetIssuerClaimTest(@ArquillianResource URL url) {
+        final JsonWebToken jwt = new JwtHelper(keyTool, "").generateProperSignedJwt();
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtIssuerValueTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+                .subject(SUBJECT)
+                .userPrincipalName(SUBJECT)
+                .issuedAtTime(now.getEpochSecond())
+                .expirationTime(later.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        final JsonWebToken jwt = JwtHelper.generateProperSignedJwtWithClaims(keyTool, jwtClaims);
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+                .issuedAtTime(now.getEpochSecond())
+                .expirationTime(later.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
+
+        final JsonWebToken jwt = JwtHelper.generateProperSignedJwtWithClaims(keyTool, jwtClaims);
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+                .userPrincipalName(SUBJECT)
+                .issuedAtTime(now.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
+
+        final JsonWebToken jwt = JwtHelper.generateProperSignedJwtWithClaims(keyTool, jwtClaims);
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+                .userPrincipalName(SUBJECT)
+                .expirationTime(later.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
+
+        final JsonWebToken jwt = JwtHelper.generateProperSignedJwtWithClaims(keyTool, jwtClaims);
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+                .issuedAtTime(now.getEpochSecond())
+                .expirationTime(later.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
+
+        final JsonWebToken jwt = JwtHelper.generateProperSignedJwtWithClaims(keyTool, jwtClaims);
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+                .userPrincipalName(SUBJECT)
+                .issuedAtTime(now.getEpochSecond())
+                .expirationTime(later.getEpochSecond())
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
+
+        final JsonWebToken jwt = JwtHelper.generateProperSignedJwtWithClaims(keyTool, jwtClaims);
+
+        RestAssured.given()
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: static import

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/package-info.java [hunk 11]
```diff
@@ -0,0 +1,4 @@
+/**
+ * Set of tests verifying specific specification compatibility related functions.
+ */
+package org.jboss.eap.qe.microprofile.jwt.compatibility;
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/package-info.java]: Create an issue for this so we can have and task to document other packages.

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/CorruptedJwtTest.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+                .header("Authorization", "Bearer " + corruptedJsonJWT.getRawValue())
+                .when().get(url.toExternalForm() + "secured-endpoint")
+                .then()
+                .statusCode(401);
+    }
+
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/CorruptedJwtTest.java]: unnecessary new line

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtIssuerValueTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    public void supplyJwtWithIncorrectIssuerClaimTest(@ArquillianResource URL url) {
+        final JsonWebToken jwt = new JwtHelper(keyTool, "fooqux42").generateProperSignedJwt();
+
+        RestAssured.given()
+                .header("Authorization", "Bearer " + jwt.getRawValue())
+                .when().get(url.toExternalForm() + "secured-endpoint")
+                .then()
+                .statusCode(401);
+    }
+
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtIssuerValueTest.java]: unnecessary new line

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+                .header("Authorization", "Bearer " + jwt.getRawValue())
+                .when().get(url.toExternalForm() + "secured-endpoint")
+                .then()
+                .statusCode(401);
+    }
+
```
Comment (istraka) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: unnecessary new line

**Comment by istraka** [comment]
LGTM, just few minor issues.

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/CorruptedJwtTest.java (hunk 12 above)
Comment (honza-kasik) [reply microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/CorruptedJwtTest.java]: Thanks - I try to always leave an empty line between class/method brackets. Though I can surely remove the line.

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+        System.out.println(jwtClaims.toJson().toString());
+
+        final JsonWebToken jwt = JwtHelper.generateProperSignedJwtWithClaims(keyTool, jwtClaims);
+
+        given().header("Authorization", "Bearer " + jwt.getRawValue())
+                .when().get(url.toExternalForm() + "secured-endpoint")
+                .then()
+                //.statusCode(401)
+                //.and()
```
Comment (mnovak1) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: I guess those can be removed.

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/package-info.java (hunk 11 above)
Comment (mnovak1) [reply microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/package-info.java]: I think I'm missing here something. Are we required to add `package-info.java ` when providing JavaDoc for test plan generator?

Comment (honza-kasik) [reply microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/package-info.java]: AFAIK no, we are not, this is just for future reference and my own initiative.

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java (hunk 15 above)
Comment (honza-kasik) [reply microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: Fixed, thanks.

**Comment by honza-kasik** [comment]
@mnovak1 All fixed, thanks.

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/package-info.java (hunk 11 above)
Comment (mnovak1) [reply microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/package-info.java]: ok, I guess there should be issue added ;-)

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+                .jwtId(UUID.randomUUID().toString())
+                .audience("microprofile-jwt-testsuite")
+                .issuer(ISSUER)
+                .userPrincipalName(SUBJECT)
+                .issuedAtTime(now.getEpochSecond())
+                .expirationTime(later.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
```
Comment (mnovak1) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: could you remove `System.out.println(...)`, please?

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+                .jwtId(UUID.randomUUID().toString())
+                .audience("microprofile-jwt-testsuite")
+                .issuer(ISSUER)
+                .subject(SUBJECT)
+                .userPrincipalName(SUBJECT)
+                .issuedAtTime(now.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
```
Comment (mnovak1) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: Remove

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+                .jwtId(UUID.randomUUID().toString())
+                .audience("microprofile-jwt-testsuite")
+                .issuer(ISSUER)
+                .subject(SUBJECT)
+                .userPrincipalName(SUBJECT)
+                .expirationTime(later.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
```
Comment (mnovak1) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: Remove

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+                .jwtId(UUID.randomUUID().toString())
+                .audience("microprofile-jwt-testsuite")
+                .issuer(ISSUER)
+                .subject(SUBJECT)
+                .issuedAtTime(now.getEpochSecond())
+                .expirationTime(later.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
```
Comment (mnovak1) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: Remove

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+                .jwtId(UUID.randomUUID().toString())
+                .audience("microprofile-jwt-testsuite")
+                .issuer(ISSUER)
+                .subject(SUBJECT)
+                .userPrincipalName(SUBJECT)
+                .issuedAtTime(now.getEpochSecond())
+                .expirationTime(later.getEpochSecond())
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
```
Comment (mnovak1) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: Remove

File: microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java [hunk 21]
```diff
@@ -0,0 +1,289 @@
...
+                .audience("microprofile-jwt-testsuite")
+                .issuer(ISSUER)
+                .subject(SUBJECT)
+                .userPrincipalName(SUBJECT)
+                .issuedAtTime(now.getEpochSecond())
+                .expirationTime(later.getEpochSecond())
+                .groups(new HashSet<>(Arrays.asList("group1", "group2")))
+                .build();
+
+        System.out.println(jwtClaims.toJson().toString());
```
Comment (mnovak1) [inline microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/compatibility/JwtMinimalClaimSetTest.java]: Remove

**Comment by mnovak1** [comment]
@honza-kasik found System.out.println() to be remove otherwise LGTM.

**Comment by honza-kasik** [comment]
@mnovak1 Thanks! Squashed and force pushed.
